components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,859 @@
+//!
+//! ## Apply (`Apply::apply_commit_merge`)
+//!
+//! ```text
+//!                 (5) CommitMergeResult
+//!        +------------------------------------+
+//!        |                                    |
+//! +------+-------+   (1) CommitMerge   +------v------+  (6) MergeResult
+//! | target apply <---------------------+ target peer +----------+
+//! +---^--+-------+                     +-----^v------+          |
+//!     |  +-------------------------------(redirect)             |
+//!     |                                       |                 |
+//!     | (4) SourceReady                       | (2) CatchUpLogs |
+//!     |                                       |                 |
+//! +---+----------+  (3) LogsUpToDate   +------v------+          |
+//! | source apply <---------------------+ source peer <----------+
+//! +--------------+                     +-------------+
+//! ```
+//!
+//! At first, target region will not apply the `CommitMerge` command. Instead
+//! the apply progress will be paused and it redirects the log entries from
+//! source region, as a `CatchUpLogs` message, to the local source region peer
+//! (step 2). When the source region peer has applied all logs up to the prior
+//! `PrepareMerge` command, it will send its tablet to the target region (step
+//! 4). Here we use a temporary channel instead of directly sending message
+//! between apply FSMs like in v1.
+//!
+//! ## On Apply Result (`Peer::on_apply_res_commit_merge`)
+//!
+//! Update the target peer states and send a `MergeResult` message to source
+//! peer to destroy it.
+
+use std::{
+    any::Any,
+    cmp::{self, Ordering},
+    fmt::{self, Debug},
+    path::PathBuf,
+};
+
+use engine_traits::{KvEngine, RaftEngine, RaftLogBatch, TabletContext, TabletRegistry};
+use futures::channel::oneshot;
+use kvproto::{
+    metapb::{self, Region},
+    raft_cmdpb::{AdminCmdType, AdminRequest, AdminResponse, CommitMergeRequest, RaftCmdRequest},
+    raft_serverpb::{MergeState, PeerState, RegionLocalState},
+};
+use protobuf::Message;
+use raft::{GetEntriesContext, Storage, INVALID_ID, NO_LIMIT};
+use raftstore::{
+    coprocessor::RegionChangeReason,
+    store::{
+        fsm::new_admin_request, metrics::PEER_ADMIN_CMD_COUNTER, util, MergeResultKind,
+        ProposalContext, Transport, WriteTask, RAFT_INIT_LOG_INDEX,
+    },
+    Error, Result,
+};
+use slog::{debug, error, info};
+use tikv_util::{
+    box_err, slog_panic,
+    store::{find_peer, region_on_same_stores},
+};
+
+use crate::{
+    batch::StoreContext,
+    fsm::ApplyResReporter,
+    operation::{AdminCmdResult, SharedReadTablet},
+    raft::{Apply, Peer},
+    router::{ApplyTask, PeerMsg, PeerTick},
+};
+
+#[derive(Debug)]
+pub struct CommitMergeResult {
+    pub index: u64,
+    region: Region,
+    source_index: u64,
+    source: Region,
+    tablet: Box<dyn Any + Send + Sync>,
+}
+
+struct SourceReady {
+    index: u64,
+    tablet: Box<dyn Any + Send + Sync>,
```

Comment:
@BusyJay Could you elaborate a bit on the benefit of using a checkpoint here? I didn't think of how using a tablet can make crash recovery difficult.

Reply:
Because now both source and target will be flushed and then merge to a new directory, so whether using a checkpoint will not affect recovery.

However, in the long term, if we want to avoid jitters caused by merge, we will need to avoid too many flushes here. For example, only flush in the source peer (as source is usually inactive). In this case, if using a checkpoint, source peer can be destroyed following the same process as v1, otherwise it will be hard to choose a correct timing. On the other hand, using a checkpoint can also make merging quicker as target region can skip LogsUpToDate if it detects a source checkpoint already exists (in most cases this is true for target region leader).

Reply:
What if source region is recreated and merged again. Will there be conflict between two different source checkpoints? I'm currently using `merge.get_commit()` as the checkpoint path index.